core/client/fs/src/main/java/alluxio/client/file/FileSystem.java

```diff
@@ -89,11 +116,115 @@ public static FileSystem get(FileSystemContext context) {
       return BaseFileSystem.create(context);
     }
 
-    public static FileSystem get(ClientContext ctx) {
-      return get(FileSystemContext.create(ctx));
+    public static FileSystem create(ClientContext ctx) {
+      return create(FileSystemContext.create(ctx));
     }
   }
 
+  /**
+   * A cache for storing {@link FileSystem} clients. This should only be used by the Factory class.
+   */
+  class Cache {
+
+    private final Object mFileSystemCacheLock = new Object();
+
+    @GuardedBy("mFileSystemCacheLock")
+    final ConcurrentHashMap<FileSystemKey, FileSystem> mCacheMap = new ConcurrentHashMap<>();
+
+    public Cache() { }
+
+    /**
+     * Gets a {@link FileSystem} from the cache. If there is none, one is created, inserted into
+     * the cache, and returned back to the user.
+     *
+     * @param key the key to retrieve a {@link FileSystem}
+     * @return the {@link FileSystem} associated with the key
+     */
+    public FileSystem get(FileSystemKey key) {
+      synchronized (mFileSystemCacheLock) {
+        return mCacheMap.computeIfAbsent(key,
+            (fileSystemKey) -> Factory.create(ClientContext.create(key.mSubject, key.mConf)));
+      }
+    }
+
+    /**
+     * Removes the client with the given key from the cache. Returns the client back to the user.
+     *
+     * @param key the client key to remove
+     * @return The removed context or null if there is no client associated with the key
+     */
+    public FileSystem remove(FileSystemKey key) {
+      synchronized (mFileSystemCacheLock) {
+        return mCacheMap.remove(key);
+      }
+    }
+
+    /**
+     * Closes and removes all {@link FileSystem} from the cache.
+     */
+    @VisibleForTesting
+    public void purge() {
+      synchronized (mFileSystemCacheLock) {
+        mCacheMap.forEach(((fileSystemKey, fileSystem) -> {
+          try {
+            fileSystem.close();
```

Comment:
Can this close be expensive? If so we should do it outside of the lock. Purge is only used in unit tests?

Reply:
`purge` is only utilized in tests. That's why it's marked `@VisibleForTesting`. It should never be used in production code. I don't see a better way to close and remove the contexts in the cache.

Reply:
One way is to clear the map in the cache after collecting all the objects, and then closing the objects outside of the lock. Or is it important that the file systems are closed before they are removed from the cache?

Reply:
I do see that we could reduce the amount of time we are locking. However I because this is just used for testing it provides no real benefit because this won't be called concurrently with operations on the cache.